File: Modes/BrowserMode.swift
/// Controller profile active when any Chromium-family browser is frontmost.
///
/// Covers Comet, Chrome, Arc, Edge, Brave, and Vivaldi — all share identical
/// keyboard shortcuts so a single mode works across all of them.
///
/// Tab navigation is the primary purpose of this mode. The consistent L1/R1 =
/// prev/next-tab mapping makes it feel like a channel remote.
///
/// ## Combo
/// `A + B` = Bookmark page. Bookmarking is behind a combo to prevent accidental
/// triggers: `A` alone opens a new tab, but `B` pressed while `A` is held
/// triggers the bookmark shortcut instead of going back.
struct BrowserMode: AppMode {

    let id           = "browser"
    let displayName  = "Browser"
    let sfSymbolName = "globe"

    var buttonMap: [GamepadButton: Action] {
        [
            // ── Face buttons ──────────────────────────────────────────────

            .a: .keystroke(
                .t,
                modifiers: .maskCommand,
                id: "browser-new-tab",
                name: "New Tab"
            ),

            // B alone = navigate back.
            // B while A is held = bookmark (handled in comboMap).
            .b: .keystroke(
                .leftBracket,
                modifiers: .maskCommand,
                id: "browser-go-back",
                name: "Go Back"
            ),

            .x: .keystroke(
                .w,
                modifiers: .maskCommand,
                id: "browser-close-tab",
                name: "Close Tab"
            ),

            // Reopen the most recently closed tab.
            .y: .keystroke(
                .t,
                modifiers: [.maskCommand, .maskShift],
                id: "browser-reopen-tab",
                name: "Reopen Closed Tab"
            ),

            // ── D-Pad ─────────────────────────────────────────────────────

            .dpadLeft: .keystroke(
                .leftBracket,
                modifiers: .maskCommand,
                id: "browser-back",
                name: "Page Back"
            ),

            .dpadRight: .keystroke(
                .rightBracket,
                modifiers: .maskCommand,
                id: "browser-forward",
                name: "Page Forward"
            ),

            .dpadUp: .keystroke(
                .pageUp,
                id: "browser-scroll-up",
                name: "Scroll Up"
            ),

            .dpadDown: .keystroke(
                .pageDown,
                id: "browser-scroll-down",
                name: "Scroll Down"
            ),

            // ── Shoulder buttons (tab cycling) ────────────────────────────

            .leftShoulder: .keystroke(
                .leftBracket,
                modifiers: [.maskCommand, .maskShift],
                id: "browser-prev-tab",
                name: "Previous Tab"
            ),

            .rightShoulder: .keystroke(
                .rightBracket,
                modifiers: [.maskCommand, .maskShift],
                id: "browser-next-tab",
                name: "Next Tab"
            ),

            // ── Triggers ─────────────────────────────────────────────────

            .leftTrigger: .keystroke(
                .n,
                modifiers: .maskCommand,
                id: "browser-new-window",
                name: "New Window"
            ),

            // Cmd+Shift+A = Search tabs in Chrome/Comet/Arc.
            .rightTrigger: .keystroke(
                .a,
                modifiers: [.maskCommand, .maskShift],
                id: "browser-tab-search",
                name: "Search Tabs"
            ),

            // ── Menu buttons ──────────────────────────────────────────────

            // Focus the address bar so she can type a URL.
            .start: .keystroke(
                .l,
                modifiers: .maskCommand,
                id: "browser-focus-address-bar",
                name: "Focus Address Bar"
            ),

            // Zoom in / out could live here — leaving unassigned for now.
        ]
    }

    // MARK: - Combos

    var comboMap: [ButtonCombo: Action] {
        [
            // A + B together = Bookmark.
            // A alone fires "new tab"; B alone fires "go back".
            // When B is pressed while A is held, the bookmark shortcut fires
            // instead of "go back", because the combo takes priority.
            ButtonCombo(.a, .b): .keystroke(
                .d,
                modifiers: .maskCommand,
                id: "browser-bookmark",
                name: "Bookmark Page"
            ),
        ]
    }
}

File: pocket_study_remote/modes/browser_mode.py
from __future__ import annotations

from ..core.action import Action, keystroke
from ..core.app_mode import AppMode
from ..core.button_combo import ButtonCombo
from ..core.gamepad_button import GamepadButton


class BrowserMode(AppMode):
    """
    Active when any web browser is frontmost.

    Covers Chrome, Safari, Firefox, Arc, Edge, Brave, Opera, Vivaldi,
    Orion, Zen Browser, and Comet. All share standard macOS browser
    shortcuts (Cmd+T, Cmd+W, Cmd+[, Cmd+Shift+[, etc.).

    Shortcuts:
    - A: New Tab (Cmd+T)
    - B: Go Back (Cmd+[)
    - X: Close Tab (Cmd+W)
    - Y: Reopen Closed Tab (Cmd+Shift+T)
    - A+B (combo): Bookmark Page (Cmd+D)
    - D-Pad Left/Right: Back/Forward
    - D-Pad Up/Down: Scroll Page
    - L1/R1: Previous/Next Tab
    - L2: New Window
    - R2: Search Tabs
    - Start: Focus Address Bar
    """

    id             = "browser"
    display_name   = "Browser"
    sf_symbol_name = "globe"

    @property
    def button_map(self) -> dict[GamepadButton, Action]:
        return {
            # ── Face buttons ──────────────────────────────────────────
            GamepadButton.A: keystroke(
                "t", ["cmd"],
                id="browser-new-tab",
                name="New Tab",
            ),
            # B alone = back. B while A is held = bookmark (see combo_map).
            GamepadButton.B: keystroke(
                "left_bracket", ["cmd"],
                id="browser-go-back",
                name="Go Back",
            ),
            GamepadButton.X: keystroke(
                "w", ["cmd"],
                id="browser-close-tab",
                name="Close Tab",
            ),
            GamepadButton.Y: keystroke(
                "t", ["cmd", "shift"],
                id="browser-reopen-tab",
                name="Reopen Closed Tab",
            ),

            # ── D-pad ─────────────────────────────────────────────────
            GamepadButton.DPAD_LEFT: keystroke(
                "left_bracket", ["cmd"],
                id="browser-back",
                name="Page Back",
            ),
            GamepadButton.DPAD_RIGHT: keystroke(
                "right_bracket", ["cmd"],
                id="browser-forward",
                name="Page Forward",
            ),
            GamepadButton.DPAD_UP: keystroke(
                "page_up",
                id="browser-scroll-up",
                name="Scroll Up",
            ),
            GamepadButton.DPAD_DOWN: keystroke(
                "page_down",
                id="browser-scroll-down",
                name="Scroll Down",
            ),

            # ── Shoulder buttons (tab cycling) ────────────────────────
            GamepadButton.LEFT_SHOULDER: keystroke(
                "left_bracket", ["cmd", "shift"],
                id="browser-prev-tab",
                name="Previous Tab",
            ),
            GamepadButton.RIGHT_SHOULDER: keystroke(
                "right_bracket", ["cmd", "shift"],
                id="browser-next-tab",
                name="Next Tab",
            ),

            # ── Triggers ─────────────────────────────────────────────
            GamepadButton.LEFT_TRIGGER: keystroke(
                "n", ["cmd"],
                id="browser-new-window",
                name="New Window",
            ),
            # Cmd+Shift+A = tab search in Chrome / Comet / Arc
            GamepadButton.RIGHT_TRIGGER: keystroke(
                "a", ["cmd", "shift"],
                id="browser-tab-search",
                name="Search Tabs",
            ),

            # ── Menu buttons ──────────────────────────────────────────
            GamepadButton.START: keystroke(
                "l", ["cmd"],
                id="browser-focus-address-bar",
                name="Focus Address Bar",
            ),
            # SELECT unassigned in V1
        }

    @property
    def combo_map(self) -> dict[ButtonCombo, Action]:
        return {
            # A + B together = Bookmark (Cmd+D)
            ButtonCombo(GamepadButton.A, GamepadButton.B): keystroke(
                "d", ["cmd"],
                id="browser-bookmark",
                name="Bookmark Page",
            ),
        }
